elk/extraction/extraction.py

```diff
@@ -148,24 +135,69 @@ def extract_hiddens(
             )
             for layer_idx in layer_indices
         }
+        lm_preds = torch.empty(
+            num_variants,
+            2,  # contrast pair
+            device=device,
+            dtype=torch.float32,
+        )
         text_inputs = []
 
         # Iterate over variants
         for i, record in enumerate(example["prompts"]):
             variant_inputs = []
 
             # Iterate over answers
-            for j in range(2):
-                text = record[j]["text"]
+            for j, choice in enumerate(record):
+                text = choice["text"]
                 variant_inputs.append(text)
 
+                # TODO: Do something smarter than "rindex" here. Really we want to
+                # get the span of the answer directly from Jinja, but that doesn't
+                # seem possible. This approach may fail for complex templates.
+                answer_start = text.rindex(choice["answer"])
+
+                # Only feed question, not the answer, to the encoder for enc-dec models
+                if model.config.is_encoder_decoder:
+                    # TODO: Maybe make this more generic for complex templates?
+                    text = text[:answer_start].rstrip()
+                    target = choice["answer"]
+                else:
+                    target = None
+
                 inputs = tokenizer(
                     text,
+                    return_offsets_mapping=True,
                     return_tensors="pt",
+                    text_target=target,  # type: ignore[arg-type]
                     truncation=True,
-                ).to(device)
+                )
+
+                # The offset_mapping is a sorted list of (start, end) tuples. We locate
+                # the start of the answer in the tokenized sequence with binary search.
+                offsets = inputs.pop("offset_mapping").squeeze().tolist()
+                inputs = inputs.to(device)
+
+                # Run the forward pass
                 outputs = model(**inputs, output_hidden_states=True)
 
+                # Compute the log probability of the answer tokens if available
+                if type(outputs).__name__.startswith("CausalLMOutput"):
+                    start, end = convert_span(
+                        offsets, (answer_start, answer_start + len(choice["answer"]))
+                    )
+                    log_p = outputs.logits[..., start - 1 : end - 1, :].log_softmax(
+                        dim=-1
+                    )
+                    tokens = inputs.input_ids[..., start:end, None]
+                    lm_preds[i, j] = log_p.gather(-1, tokens).sum()
```

Comment:
This part is still a bit confusing to me, and I'll have a look at it in detail tomorrow. Apart from that, the code seems to work correctly and everything else lgtm.

Reply:
Is it still confusing